extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java

```diff
@@ -158,174 +131,73 @@ public List<WritableFrameChannel> outputChannels()
-        2.2. If they don't match, then we have reached a partition boundary.
-             In this case, we update the value for lastPartitionIndex.
-     3. End of Input: If the input channel is finished, any remaining rows in rowsToProcess are processed.
-
-     *Illustration of Row Comparison step*
-
-     Let's say we have window_function() OVER (PARTITION BY A ORDER BY B) in our query, and we get 3 frames in the input channel to process.
-
-     Frame 1
-     A, B
-     1, 2
-     1, 3
-     2, 1 --> PARTITION BY key (column A) changed from 1 to 2.
-     2, 2
-
-     Frame 2
-     A, B
-     3, 1 --> PARTITION BY key (column A) changed from 2 to 3.
-     3, 2
-     3, 3
-     3, 4
-
-     Frame 3
-     A, B
-     3, 5
-     3, 6
-     4, 1 --> PARTITION BY key (column A) changed from 3 to 4.
-     4, 2
-
-     *Why batching?*
-     We batch multiple PARTITION BY keys for processing together to avoid the overhead of creating different RACs for each PARTITION BY keys, as that would be unnecessary in scenarios where we have a large number of PARTITION BY keys, but each key having a single row.
-
-     *Future thoughts: https://github.com/apache/druid/issues/16126*
-     Current approach with R&C and operators materialize a single R&C for processing. In case of data with low cardinality a single R&C might be too big to consume. Same for the case of empty OVER() clause.
-     Most of the window operations like SUM(), RANK(), RANGE() etc. can be made with 2 passes of the data. We might think to reimplement them in the MSQ way so that we do not have to materialize so much data.
-     */
-
     // If there are rows pending flush, flush them and run again before processing any more rows.
     if (frameHasRowsPendingFlush()) {
       flushAllRowsAndCols();
       return ReturnOrAwait.runAgain();
     }
 
-    if (partitionColumnNames.isEmpty()) {
-      // Scenario 1: Query has atleast one window function with an OVER() clause without a PARTITION BY.
-      if (inputChannel.canRead()) {
-        final Frame frame = inputChannel.read();
-        convertRowFrameToRowsAndColumns(frame);
-        return ReturnOrAwait.runAgain();
-      }
+    if (inputChannel.canRead()) {
+      final Frame frame = inputChannel.read();
+      LazilyDecoratedRowsAndColumns ldrc = convertRowFrameToRowsAndColumns(frame);
+      frameRowsAndColsBuilder.add(ldrc);
 
-      if (inputChannel.isFinished()) {
-        // If no rows are flushed yet, process all rows.
-        if (rowId.get() == 0) {
-          runAllOpsOnMultipleRac(frameRowsAndCols);
+      if (needToProcessBatch()) {
+        runAllOpsOnBatch();
+        try {
+          flushAllRowsAndCols();
         }
-
-        // If there are still rows pending after operations, run again.
-        if (frameHasRowsPendingFlush()) {
-          return ReturnOrAwait.runAgain();
+        catch (IOException e) {
+          throw new RuntimeException(e);
```

Comment:
why the need to catch this? 
at line 136 it was okay to throw - why not okay here?

Reply:
Ah right, have made the change.